Vwitter/View Controllers/SearchUsersViewController.m

```diff
@@ -0,0 +1,92 @@
+//
+//  SearchUsersViewController.m
+//  Vwitter
+//
+//  Created by Christina Li on 7/8/22.
+//
+
+#import "SearchUsersViewController.h"
+#import "UserCell.h"
+#import <Parse/Parse.h>
+
+@interface SearchUsersViewController () <UITableViewDataSource, UITableViewDelegate, UISearchBarDelegate>
+@property (weak, nonatomic) IBOutlet UITableView *tableView;
+
+@property (strong, nonatomic) NSMutableArray *arrayOfUsers;
+@property (strong, nonatomic) NSArray *filteredData;
+@property (weak, nonatomic) IBOutlet UISearchBar *searchBar;
+// make a property that stores the people that the user is following
+
+@end
+
+@implementation SearchUsersViewController
+
+- (void)viewDidLoad {
+    [super viewDidLoad];
+    // Do any additional setup after loading the view.
+    [self loadData];
+    
+    self.tableView.dataSource = self;
+    self.tableView.delegate = self;
+    self.searchBar.delegate = self;
+    self.filteredData = self.arrayOfUsers;
+    
+}
+
+- (void)loadData {
+    PFQuery *userQuery = [PFUser query];
+    userQuery.limit = 20;
+    __weak typeof(self) weakSelf = self;
+    [userQuery findObjectsInBackgroundWithBlock:^(NSArray<PFUser *> * _Nullable users, NSError * _Nullable error) {
+        typeof(self) strongSelf = weakSelf;
+        if (!strongSelf) {
+            NSLog(@"I got killed!");
+            return;
+        }
+        if (users) {
+            NSLog(@"😎😎😎 Successfully loaded search users timeline");
+            strongSelf.arrayOfUsers = users.mutableCopy;
+            strongSelf.filteredData = strongSelf.arrayOfUsers;
+            [strongSelf.tableView reloadData];
+            
+        }
+        else {
+            NSLog(@"😫😫😫 Error getting home timeline: %@", error.localizedDescription);
+        }
+    }];
+}
+
+- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
+    return self.filteredData.count;
+}
+
+- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
+    UserCell *cell = [tableView dequeueReusableCellWithIdentifier:@"UserCell" forIndexPath:indexPath];
+//    if this userid is in the followingSet, then do stuff inside the usercell
+
+    cell.user = self.filteredData[indexPath.row];
+
+    return cell;
+}
+
+- (void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText {
+    
+    if (searchText.length != 0) {
+        
+        NSPredicate *predicate = [NSPredicate predicateWithBlock:^BOOL(PFUser *evaluatedObject, NSDictionary *bindings) {
+            return [evaluatedObject.username containsString:searchText];
+        }];
+        self.filteredData = [self.arrayOfUsers filteredArrayUsingPredicate:predicate];
```

Comment:
It is great to see `NSPredicate`, it's been a long time! 

It is much better to do this kind of logic on the backend. You can define a Parse Cloud function, something like `"search"`, which will take some input string parameter and run the actual search. Any time the search bar text changes, you can call that endpoint. 

One potential problem will be, users might be rapidly typing, and so the user might have changed text 5 times by the time the first network request comes back. For this, you can use a "currentRequestCount" variable so that you know if the search should be ignored because a new one was already executed

first, you would create a property of the view controller, an integer like this

```
@property (nonatomic, readwrite) int requestCount;
```

Then inside of this `searchBar:textDidChange:` method, you can check it like this

```
- (void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText {
    const int currentRequestCount = self.requestCount++;
    __weak typeof(self) weakSelf = self;
    [PFCloud callFunctionInBackground:@"search" withParameters:@{@"search_text" : searchText} block:^(id object, NSError *error) {
        typeof(self) strongSelf = weakSelf;
        if (!strongSelf) {
            NSLog(@"%@ died!", NSStringFromClass(self.class));
            return;
        }
        if (strongSelf.requestCount > currentRequestCount) {
            NSLog(@"Cancelled search because a new search is already in progress");
            return;
        }
        // handle the search results
    }];
}
```

That way, if the user quickly typed and 5 search requests have gone out, the client won't waste time processing requests 1 to 4.